doc/examples/server_selection.rst

```diff
@@ -0,0 +1,114 @@
+The server selector function should accept a list of
+:class:`~pymongo.server_description.ServerDescription` objects and return a
+list of server descriptions that are suitable for the read or write operation.
+A server selector must not create or modify
+:class:`~pymongo.server_description.ServerDescription` objects, and must return
+the selected instances unchanged.
+
+In this example, we write a server selector that prioritizes servers running on
+``localhost``. This can be desirable when using a sharded cluster with multiple
+``mongos``, as locally run queries are likely to see lower latency and higher
+throughput. Please note, however, that it is highly dependent on the
+application if preferring ``localhost`` is beneficial or not.
+
+In addition to comparing the hostname with ``localhost``, our server selector
+function accounts for the edge case when no servers are running on
+``localhost``. In this case, we allow the default server selection logic to
+prevail by passing through the received server description list unchanged.
+Failure to do this would render the client unable to communicate with MongoDB
+in the event that no servers were running on ``localhost``.
+
+
+The described server selection logic is implemented in the following server
+selector function:
+
+
+.. doctest::
+
+   >>> def server_selector(server_descriptions):
+   ...     servers = [
+   ...         server for server in server_descriptions
+   ...         if server.address[0] == 'localhost'
+   ...     ]
+   ...     if not servers:
+   ...         return server_descriptions
+   ...     return servers
+
+
+
+Finally, we can create a :class:`~pymongo.MongoClient` instance with this
+server selector.
+
+
+.. doctest::
+
+   >>> client = MongoClient(server_selector=server_selector)
+
+
+
+Server Selection Process
+------------------------
+
+This section dives deeper into the server selection process for reads and
+writes. Let's first consider the case of **writes**. Assuming it is connected to a
+reachable database, the driver performs the following operations (in order)
+during the selection process:
+
+* Select all writeable servers from the list of known hosts. For a replica set
+  this is the primary, while for a sharded cluster this can be an arbitrary
+  number of ``mongos``
+
+* Apply the user-defined server selector function. Note that the custom server
+  selector is **not** called if there are no servers left from the previous
+  filtering stage.
+
+* Apply the ``localThresholdMS`` setting to the list of remaining hosts. This
+  whittles the host list down to only contain servers whose latency is at most
+  ``localThresholdMS`` milliseconds higher than the lowest observed latency.
+
+* Select a server at random from the remaining host list. The desired
+  operation is then performed against the selected server.
+
+
+For the case of **reads** the process is identical except for the first step.
```

Comment:
"*In* the case of" is much more popular than "for": https://books.google.com/ngrams/graph?content=for+the+case+of%2Cin+the+case+of&year_start=1800&year_end=2000&corpus=15&smoothing=3&share=&direct_url=t1%3B%2Cfor%20the%20case%20of%3B%2Cc0%3B.t1%3B%2Cin%20the%20case%20of%3B%2Cc0